src/nbl/video/IGPUCommandBuffer.cpp

```diff
@@ -0,0 +1,951 @@
+
+    if (!m_cmdpool->emplace<IGPUCommandPool::CPipelineBarrierCmd>(m_segmentListHeadItr, m_segmentListTail, bufferMemoryBarrierCount, bufferResources, imageMemoryBarrierCount, imageResources))
+        return false;
+
+    pipelineBarrier_impl(srcStageMask, dstStageMask, dependencyFlags, memoryBarrierCount, pMemoryBarriers, bufferMemoryBarrierCount, pBufferMemoryBarriers, imageMemoryBarrierCount, pImageMemoryBarriers);
+
+    return true;
+}
+
+bool IGPUCommandBuffer::bindDescriptorSets(asset::E_PIPELINE_BIND_POINT pipelineBindPoint, const pipeline_layout_t* layout, uint32_t firstSet, const uint32_t descriptorSetCount,
+    const descriptor_set_t* const* const pDescriptorSets, const uint32_t dynamicOffsetCount, const uint32_t* dynamicOffsets)
+{
+    if (!this->isCompatibleDevicewise(layout))
+        return false;
+
+    if (layout->getAPIType() != getAPIType())
+        return false;
+
+    for (uint32_t i = 0u; i < descriptorSetCount; ++i)
+    {
+        if (pDescriptorSets[i])
+        {
+            if (!this->isCompatibleDevicewise(pDescriptorSets[i]))
+            {
+                m_logger.log("IGPUCommandBuffer::bindDescriptorSets failed, pDescriptorSets[%d] was not created by the same ILogicalDevice as the commandbuffer!", system::ILogger::ELL_ERROR, i);
+                return false;
+            }
+            if (!pDescriptorSets[i]->getLayout()->isIdenticallyDefined(layout->getDescriptorSetLayout(firstSet + i)))
+            {
+                m_logger.log("IGPUCommandBuffer::bindDescriptorSets failed, pDescriptorSets[%d] not identically defined as layout's %dth descriptor layout!", system::ILogger::ELL_ERROR, i, firstSet + i);
+                return false;
+            }
+        }
+    }
+
+    core::smart_refctd_ptr<const video::IGPUDescriptorSet> descriptorSets_refctd[IGPUPipelineLayout::DESCRIPTOR_SET_COUNT] = { nullptr };
+    for (auto i = 0; i < descriptorSetCount; ++i)
+    {
+        assert(i < IGPUPipelineLayout::DESCRIPTOR_SET_COUNT);
+        descriptorSets_refctd[i] = core::smart_refctd_ptr<const video::IGPUDescriptorSet>(pDescriptorSets[i]);
+    }
+
+    if (!m_cmdpool->emplace<IGPUCommandPool::CBindDescriptorSetsCmd>(m_segmentListHeadItr, m_segmentListTail, core::smart_refctd_ptr<const IGPUPipelineLayout>(layout), descriptorSetCount, descriptorSets_refctd))
+        return false;
+
+    for (uint32_t i = 0u; i < descriptorSetCount; ++i)
+    {
+        if (pDescriptorSets[i] && !pDescriptorSets[i]->getLayout()->canUpdateAfterBind())
+        {
+            const auto currentVersion = pDescriptorSets[i]->getVersion();
+
+            auto found = m_boundDescriptorSetsRecord.find(pDescriptorSets[i]);
+
+            if (found != m_boundDescriptorSetsRecord.end())
+            {
+                if (found->second != currentVersion)
+                {
+                    m_logger.log("Descriptor set #%u version does not match that of the command buffer's bound descriptor set.", system::ILogger::ELL_ERROR, i);
```

Comment:
print pointer and debug name and better message like "was modified between two recorded bind commands since the last command buffer's beginning"

Reply:
Resolved.